### PR TITLE
A first prototype of leshan-server-demo UI v2 based on vuejs

### PR DESCRIPTION
**Pull request by sbernard31**
The idea is to rewrite the UI using modern frontend tools (see #906).
Vuejs + vuetify was used for this prototype.

The old demo UI is still available at : http://localhost:8080
The new one is available at : http://localhost:8080/v2

This is just a prototype and a lot of feature is missing but this give a good idea about code organization.
I think this is really better than what we used in the previous demo. (Maybe the first time I appreciate to write frontend code :sweat_smile:)

But using this framework changes a lot the development workflow and build way  because now the frontend code need to be build :astonished:.

Concretely that means new build steps in maven which makes the build a bit longer. (I'm using [frontend-maven-plugin](frontend-maven-plugin))

The development flow is different too.
I used **vscodium** to code frontend, opening just the `leshan-server-demo/webapp2` folder.
To test you need : 
 - to run leshan-server-demo with you preferred java IDE (eclipse for me) to serve REST API.
 - to run `yarn server` in  `leshan-server-demo/webapp2` folder and UI will be available at http://localhost:8088/v2 (the port is not the same than in production)
 - then you can modify you code in vscodium and it will be updated automatically (no need to restart `yarn server`)
 
**A capture of the new UI :**
leshan-server-demo-UIv2

### CODE REVIEW

**Comment by sbernard31** [comment]
I just pushed some fixes but there is still an issue with expansion panel (see https://github.com/vuetifyjs/vuetify/issues/13167)
To be honest I'm not sure expansion panel is the right component :thinking:

**Comment by dachaac** [comment]
Hmm... for some reason it doesn't like to start:
```
08:12:12.073 [main] ERROR org.eclipse.leshan.server.demo.LeshanServerDemo - Jetty stopped with unexpected error ...
java.lang.NullPointerException: null
	at org.eclipse.leshan.server.demo.LeshanServerDemo.createAndStartServer(LeshanServerDemo.java:560)
	at org.eclipse.leshan.server.demo.LeshanServerDemo.main(LeshanServerDemo.java:370)
```

Did maven clean install before that.

Then cleaned up all arguments from leshan-server-demo.

As like it couldn't find `webapp2`.

**Comment by sbernard31** [comment]
I just tried to reproduce it and it works for me :thinking:  : 
```
git clone git@github.com:eclipse/leshan.git leshan-test    
cd leshan-test    
git checkout demo_v2  
mvn clean install   
java -jar leshan-server-demo/target/leshan-server-demo-2.0.0-SNAPSHOT-jar-with-dependencies.jar 
```

But maybe you are not talking about `build` but rather `developement workflow`?

**Comment by sbernard31** [comment]
I did more test about development workflow and it works for me too but I guess this is because I'm using eclipse and m2e which is able to generate webapp2 folder in target/classes folder.  This slow down eclipse and maybe don't work with other idea, I will maybe try to find an alternative.

Please if you can find time,  give me more details about the step to reproduce. :pray:

**Comment by sbernard31** [comment]
I tried to improve the development workflow with my last commit and add a [README](https://github.com/eclipse/leshan/blob/demo_v2/leshan-server-demo/webapp2/README.md) with more detail about it.

HTH.

**Comment by dachaac** [comment]
What ever you did it changed things a bit ;).

As a quick notes what I have:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
```

```
$ java -version
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
```
```
$ node --version
v12.21.0
```
```
$ npm --version
6.14.11
```

Then I have (latest) IntelliJ IDEA Ultimate (2020.3.2).

Then just for this I got latest vscode:
```
Version: 1.54.1
Commit: f30a9b73e8ffc278e71575118b6bf568f04587c8
Date: 2021-03-04T22:38:50.094Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Linux x64 5.4.0-66-generic
```
(I also have webstorm here what I would normally use)

What I did now:
```
git checkout master
git fetch -v
git branch -D demo_v2
git checkout demo_v2
```
Then in IntelliJ I have maven target `clean install` in leshan's root.

Then to execute it I have leshan-server-demo application target with:
```
-cp leshan-server-demo
org.eclipse.leshan.server.demo.LeshanServerDemo
```
and working directory as leshan's root.

It seems that this target has Java 10 SDK. Tried changing this to 1.8 or to 11 without any visible effect.

Now I do get different page and it instructs one to go to /v2/...
```
If you are using a built version of leshan-server-demo and you see this page, you probably discover a leshan built issue. Please report it.
```
So something is still wrong.

Well what I next tried out was do the yarn part from command line (install + serve). And now in :8088 I do have something new -- but lists are empty not really anything you can do here. Being empty might be of course that there are no connected clients at the moment.

For sanity check -- I went for command line for Java part too.

In leshan's root:
```
java -jar leshan-server-demo/target/leshan-server-demo-2.0.0-SNAPSHOT-jar-with-dependencies.jar 
```
It behaves similar to what was in IntelliJ. Expect I don't see extra debugs.

So is the backend actually working now and I just need to get some more data there?

I do have some security configs from last time I was playing around with leshan but I suppose that UI part is not yet there.

**Comment by sbernard31** [comment]
Your comment makes me feel that my documentation is not clear enough :sweat_smile:. 

### 1. Maven build
So firstly, let's ensure that build works. 
Do you try [this](https://github.com/eclipse/leshan/pull/977#issuecomment-792606326) in **a terminal**. Does it works for you ? Using this you should not see the page : `If you are using a built version of leshan-server-demo and you see this page ... ` at http://localhost:8080/v2/. (make sure your browser does not used its cache)

I double check this on my side and it works as expected, so I'm a bit confused here. 

> It behaves similar to what was in IntelliJ. Expect I don't see extra debugs.

I don't get it ? 

### 2. Development workflow
This is normal to get the `If you are using a built version of leshan-server-demo and you see this page ... `  at http://localhost:8080/v2/.

Because you need to use `yarn serve` to get Hot Module Replacement. In the case the UI is serve at http://localhost:8088/v2/

### 3 No data

> Being empty might be of course that there are no connected clients at the moment.

Correct

> So is the backend actually working now and I just need to get some more data there?

Yep just connect a client. E.g. a Leshan client demo without security credentials.

> I do have some security configs from last time I was playing around with leshan but I suppose that UI part is not yet there.
Yep new demo doesn't have UI to set security info but the used backend is the same, there is just a new UI served on /v2 path. Meaning the old one is still usable, you can even use both at same time.
If you add security credentials with the old UI and connect a device you will see it also in the new UI.
(but the easy way is just to connect without security)

**Comment by sbernard31** [comment]
I launch a build on jenkins. I tested the[ **leshan-server-demo**]( https://ci.eclipse.org/leshan/job/leshan-hudson-test/67/artifact/), it seems to work as expected.

But I discover a kind of build issue : the build is really longer because node/yarn and dependencies are downloaded  for each build. I expected that it did this only 1 time. I guess that maybe this is a jenkins issue where workspace is deleted and should not. 
This is something I need to investigate later

**Comment by dachaac** [comment]
> I just tried to reproduce it and it works for me  :
> 
> ```
> git clone git@github.com:eclipse/leshan.git leshan-test    
> cd leshan-test    
> git checkout demo_v2  
> mvn clean install   
> java -jar leshan-server-demo/target/leshan-server-demo-2.0.0-SNAPSHOT-jar-with-dependencies.jar 
> ```
> 
> But maybe you are not talking about `build` but rather `developement workflow`?

When I run this flow exactly http://0.0.0.0:8080/ -> v2 -- I do get the new Web UI.

But if I do this from IDE like what I would normally do so that I can also debug things story seems to be different.

**Comment by dachaac** [comment]
So I assume that the setup is otherwise correct but building and running with my config in IntelliJ just does not expose the built webui thru :8080.

**Comment by dachaac** [comment]
> > It behaves similar to what was in IntelliJ. Expect I don't see extra debugs.
> 
> I don't get it ?

For some reason when I run in IntellJ lots of debug detail is visible like:
```
11:11:19.984 [main] DEBUG org.eclipse.jetty.util.component.AbstractLifeCycle - STARTED @4643ms SelectorManager@ServerConnector@48075da3{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
11:11:19.984 [main] DEBUG org.eclipse.jetty.util.component.ContainerLifeCycle - ServerConnector@48075da3{HTTP/1.1, (http/1.1)}{0.0.0.0:8080} added {acceptor-0@554e218,POJO}
11:11:19.984 [main] DEBUG org.eclipse.jetty.util.thread.QueuedThreadPool - queue acceptor-0@554e218 startThread=0
11:11:19.984 [main] INFO org.eclipse.jetty.server.AbstractConnector - Started ServerConnector@48075da3{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
11:11:19.984 [qtp658909832-36] DEBUG org.eclipse.jetty.util.thread.QueuedThreadPool - run acceptor-0@554e218 in QueuedThreadPool[qtp658909832]@27462a88{STARTED,8<=8<=200,i=1,r=12,q=0}[ReservedThreadExecutor@479ceda0{s=0/12,p=0}]
11:11:19.985 [main] DEBUG org.eclipse.jetty.util.component.AbstractLifeCycle - STARTED @4644ms ServerConnector@48075da3{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
11:11:19.985 [main] INFO org.eclipse.jetty.server.Server - Started @4644ms
11:11:19.985 [main] DEBUG org.eclipse.jetty.util.component.AbstractLifeCycle - STARTED @4644ms Server@1ed6388a{STARTED}[9.4.34.v20201102]
11:11:19.985 [main] INFO org.eclipse.leshan.server.demo.LeshanServerDemo - Web server started at http://0.0.0.0:8080/.
```

Haven't really looked why -- or if I have ticked something on some time ago. I am just happy that I see the extra details there ;)

**Comment by dachaac** [comment]
@sbernard31 so what kind of feedback do you want for this? / how you want to go forward?

**Comment by sbernard31** [comment]
> @sbernard31 so what kind of feedback do you want for this? 

Oh :sweat_smile: nothing special. As you seemed interested by the past by the subject, I just let you know that I begin to work on this. 

My current feeling is : 
- build seems more complicated/slower
- development workflow seems more complicated to understand (at least for non-frontend dev)
- but once you get it, this is really a more pleasant/efficient way to code web UI. 

My conclusion  : this could be the right move. Hoping that  vue.js popularity and good documentation will the more complicate workflow and so be not a too big obstacle for contribution.

If you see a NO-GO reason, I guess it's the good time.  

> how you want to go forward?

My idea is to integrate this new UI and also keep the old one until : 
- either the new UI is able to do like the old one and more (and so we remove the old)

Now if you want invest more time in the demo UI, I would advice to apply it only to the new UI. 

Does it make sense ?

**Comment by sbernard31** [comment]
I made some more minor fix.
I think this is in mergeable state now.

@dachaac, I will probably integrate it in master next week **unless you think this is a bad idea**. 

My next move about this new UI will be to work on the Security tab to allow to configure credentials. (unless you plan to work on it but my guess is that you haven't so much time to work on Leshan currently)

**Comment by dachaac** [comment]
@sbernard31 I don't have anything against going to this direction.

You have commented out code in the files in here -- was that intention?

I believe this will anyway be iterative work.

**Comment by dachaac** [comment]
In some projects the work flow works in a way that pre-compiled versions of webui is stored in source repo to make life easier for most of the people developing the code.

Then if you run the "full build" then it updates those "deployment" directories so that it is easy to commit updates.

**Comment by sbernard31** [comment]
> You have commented out code in the files in here -- was that intention?

I'm not sure to get which comments or files you are talking about ? :thinking: 

> In some projects the work flow works in a way that pre-compiled versions of webui is stored in source repo to make life easier for most of the people developing the code.

This is maybe something we should investigate. For now I have no clear idea about how this could work.
I must also confess that I'm bit afraid about "de-synchronization issue" between source code and built code in the repo :cold_sweat:.

**Comment by dachaac** [comment]
> > You have commented out code in the files in here -- was that intention?
> 
> I'm not sure to get which comments or files you are talking about ? 

Ok. I'll add comments for those.

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java [hunk 1]
```diff
@@ -541,19 +546,41 @@ public static void createAndStartServer(String webAddress, int webPort, String l
...
         Server server = new Server(jettyAddr);
-        WebAppContext root = new WebAppContext();
-        root.setContextPath("/");
-        root.setResourceBase(LeshanServerDemo.class.getClassLoader().getResource("webapp").toExternalForm());
-        root.setParentLoaderPriority(true);
+        /*
+         * WebAppContext root = new WebAppContext(); root.setContextPath("/");
+         * root.setResourceBase(LeshanServerDemo.class.getClassLoader().getResource("webapp").toExternalForm());
+         * root.setParentLoaderPriority(true); server.setHandler(root);
+         */
```
Comment (dachaac) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:553]: Commented code

File: leshan-server-demo/webapp2/src/App.vue [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        </v-container>
+      </v-main>
+      <!--leshan-footer /-->
+    </v-app>
+  </div>
+</template>
+
+<script>
+import LeshanNavBar from "@/components/LeshanNavBar.vue";
+//import LeshanFooter from "@/components/LeshanFooter.vue";
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/App.vue:18]: Commented code

File: leshan-server-demo/webapp2/src/App.vue [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+export default {
+  components: {
+    LeshanNavBar,
+    //LeshanFooter,
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/App.vue:23]: Commented code

File: leshan-server-demo/webapp2/src/components/ClientSetting.vue [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        "SENML_CBOR",
+        "JSON",
+      ],
+      //multiFormatList: ["TLV", "SENML_JSON", "SENML_CBOR", "JSON"],
+    };
+  },
+  computed: {
+    timeout: preference("timeout", { defaultValue: 5 }),
+    singleformat: preference("singleformat", { defaultValue: "TLV" }),
+    //multiformat: preference("multiformat", { defaultValue: "TLV" }),
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/components/ClientSetting.vue:65]: Commented code

File: leshan-server-demo/webapp2/src/components/ClientSetting.vue [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      singleFormatList: [
+        "TLV",
+        "CBOR",
+        "TEXT",
+        "OPAQUE",
+        "SENML_JSON",
+        "SENML_CBOR",
+        "JSON",
+      ],
+      //multiFormatList: ["TLV", "SENML_JSON", "SENML_CBOR", "JSON"],
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/components/ClientSetting.vue:59]: Commented code

File: leshan-server-demo/webapp2/src/plugins/axios.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
+"use strict";
+
+import Vue from 'vue';
+import axios from "axios";
+
+// Full config:  https://github.com/axios/axios#request-config
+// axios.defaults.baseURL = process.env.baseURL || process.env.apiUrl || '';
+// axios.defaults.headers.common['Authorization'] = AUTH_TOKEN;
+// axios.defaults.headers.post['Content-Type'] = 'application/x-www-form-urlencoded';
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/plugins/axios.js:9]: Commented code

File: leshan-server-demo/webapp2/src/plugins/axios.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+let config = {
+  // baseURL: process.env.baseURL || process.env.apiUrl || ""
+  // timeout: 60 * 1000, // Timeout
+  // withCredentials: true, // Check cross-site Access-Control
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/plugins/axios.js:14]: Commented code

File: leshan-server-demo/webapp2/src/plugins/vuetify.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import '@mdi/font/css/materialdesignicons.css';
+import "@fontsource/roboto"
+import Vue from "vue";
+import Vuetify from "vuetify/lib/framework";
+
+Vue.use(Vuetify);
+
+export default new Vuetify({
+  icons: {
+    iconfont: "mdi", // 'mdi' || 'mdiSvg' || 'md' || 'fa' || 'fa4' || 'faSvg'
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/plugins/vuetify.js:10]: Commented options at the end of line?

File: leshan-server-demo/webapp2/src/views/Client.vue [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+            instances.push({ id: ids[2] });
+        });
+      }
+      return instances;
+    },
+  },
+  watch: {
+    /*$route(to, from) {
+      console.log("route", to, from);
+    },*/
```
Comment (dachaac) [inline leshan-server-demo/webapp2/src/views/Client.vue:80]: Commented code

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java (hunk 1 above)
Comment (sbernard31) [reply leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/LeshanServerDemo.java:553]: This code should be added again when we will remove old version of demo.
So I will keep it (I will add a comment to explain)

File: leshan-server-demo/webapp2/src/App.vue (hunk 2 above)
Comment (sbernard31) [reply leshan-server-demo/webapp2/src/App.vue:18]: I hesitated to keep a footer but finally I think it take to much space and the about pages is enough.
I will remove all comments/files about footer.

File: leshan-server-demo/webapp2/src/components/ClientSetting.vue (hunk 4 above)
Comment (sbernard31) [reply leshan-server-demo/webapp2/src/components/ClientSetting.vue:65]: I comment  multi-node format content support as for now there is noway to write multi node with the new UI.
It should be activated later (I will keep it and add a comment)

File: leshan-server-demo/webapp2/src/plugins/axios.js (hunk 6 above)
Comment (sbernard31) [reply leshan-server-demo/webapp2/src/plugins/axios.js:9]: this is auto generated code done by `vue CLI` I don't know if we should keep it or not. I gues this is just a kind of documentation. 
I will keep it for now (at least to keep it in 1 comment), I don't know if we remove it later ?

File: leshan-server-demo/webapp2/src/plugins/axios.js (hunk 7 above)
Comment (sbernard31) [reply leshan-server-demo/webapp2/src/plugins/axios.js:14]: same as above

File: leshan-server-demo/webapp2/src/plugins/vuetify.js (hunk 8 above)
Comment (sbernard31) [reply leshan-server-demo/webapp2/src/plugins/vuetify.js:10]: same as above

File: leshan-server-demo/webapp2/src/views/Client.vue (hunk 9 above)
Comment (sbernard31) [reply leshan-server-demo/webapp2/src/views/Client.vue:80]: This is where we should handle url changes, we probably need to to that.
But not a priority so I guess I will keep it and add a TODO

**Comment by sbernard31** [comment]
@dachaac thx for all your comment and feedback. :pray: 

I will fix it and integrate it in master.

**Comment by sbernard31** [comment]
Integrated in master (squashed in 1 commit f614b269ebaf6bc61062ec010ab10d51cef933e8)

**Pull request closed without merging**